operator/builtin/input/windows/operator.go

```diff
@@ -0,0 +1,223 @@
+		InputOperator: inputOperator,
+		buffer:        NewBuffer(),
+		channel:       c.Channel,
+		maxReads:      c.MaxReads,
+		startAt:       c.StartAt,
+		pollInterval:  c.PollInterval,
+		offsets:       offsets,
+	}
+	return eventLogInput, nil
+}
+
+// NewDefaultConfig will return an event log config with default values.
+func NewDefaultConfig() operator.Builder {
+	return &EventLogConfig{
+		MaxReads: 100,
+		StartAt:  "end",
+		PollInterval: operator.Duration{
+			Duration: 5 * time.Second,
+		},
+	}
+}
+
+// EventLogInput is an operator that creates entries using the windows event log api.
+type EventLogInput struct {
+	helper.InputOperator
+	bookmark     Bookmark
+	subscription Subscription
+	buffer       Buffer
+	channel      string
+	maxReads     int
+	startAt      string
+	pollInterval operator.Duration
+	offsets      helper.Persister
+	cancel       context.CancelFunc
+	wg           *sync.WaitGroup
+}
+
+// Start will start reading events from a subscription.
+func (e *EventLogInput) Start() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	e.cancel = cancel
+	e.wg = &sync.WaitGroup{}
+
+	e.bookmark = NewBookmark()
+	offsetXML, err := e.getBookmarkOffset()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve bookmark offset: %s", err)
+	}
+
+	if err := e.bookmark.Open(offsetXML); err != nil {
+		return fmt.Errorf("failed to open bookmark: %s", err)
+	}
+
+	e.subscription = NewSubscription()
+	if err := e.subscription.Open(e.channel, e.startAt, e.bookmark); err != nil {
+		return fmt.Errorf("failed to open subscription: %s", err)
+	}
+
+	e.wg.Add(1)
+	go e.readOnInterval(ctx)
+	return nil
+}
+
+// Stop will stop reading events from a subscription.
+func (e *EventLogInput) Stop() error {
+	e.cancel()
+	e.wg.Wait()
+
+	if err := e.subscription.Close(); err != nil {
+		return fmt.Errorf("failed to close subscription: %s", err)
+	}
+
+	if err := e.bookmark.Close(); err != nil {
+		return fmt.Errorf("failed to close bookmark: %s", err)
+	}
+
+	return nil
+}
+
+// readOnInterval will read events with respect to the polling interval.
+func (e *EventLogInput) readOnInterval(ctx context.Context) {
+	defer e.wg.Done()
+
+	ticker := time.NewTicker(e.pollInterval.Raw())
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			e.read(ctx)
+		}
+	}
+}
+
+// read will read events from the subscription and update the current offset.
+func (e *EventLogInput) read(ctx context.Context) {
+	events, err := e.subscription.Read(e.maxReads)
+	if err != nil {
+		e.Errorf("Failed to read events from subscription: %s", err)
+		return
+	}
+
+	for i, event := range events {
+		e.processEvent(ctx, event)
+		if len(events) == i+1 {
+			e.updateBookmarkOffset(event)
+		}
+		event.Close()
+	}
+}
+
+// processEvent will process and send an event retrieved from windows event log.
+func (e *EventLogInput) processEvent(ctx context.Context, event Event) {
+	simpleEvent, err := event.RenderSimple(e.buffer)
```

Comment:
Looks like `simpleEvent` is only used if we fail to render the formatted event. Can we avoid rendering the simple event until we hit a failure?

Reply:
Unfortunately not. If you look closely, we use the simple event to help render the formatted event. Unlike simple event rendering, which just requires the event handle, the formatted event requires a different syscall with an additional publisher handle. That publisher handle is created with another syscall using the provider name. If we wanted to get that provider name, we'd have to do two more syscalls. However, I noticed a way around this. The simple event already has the provider name rendered. So we can eliminate 1 entire syscall by just using this instead, and we have the added benefit of having simple event already rendered if we need it.

Reply:
Aah, I see now. Thanks for the explanation